core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java

```diff
@@ -174,10 +175,13 @@ public MetaData applyChanges(MetaData oldMetaData, RoutingTable newRoutingTable)
             // Prevent set of inSyncAllocationIds to grow unboundedly. This can happen for example if we don't write to a primary
             // but repeatedly shut down nodes that have active replicas.
             // We use number_of_replicas + 1 (= possible active shard copies) to bound the inSyncAllocationIds set
+            // Only trim the set of allocation ids when it grows, otherwise we might trim too eagerly when the number
+            // of replicas was decreased while shards were unassigned.
```

Comment:
Could this hypothetically still happen if the number of replicas for an index was decreased while shards were unassigned, and simultaneously new indices were created so that the newly added check would still trim the allocation ids?